paddle/platform/profiler.h

```diff
@@ -0,0 +1,197 @@
+
+#ifdef PADDLE_WITH_CUDA
+  cudaEvent_t event() const { return event_; }
+
+  int device() const { return device_; }
+#endif
+
+  double CpuElapsedUs(const Event& e) const {
+    return (e.cpu_ns_ - cpu_ns_) / (1000.0);
+  }
+
+  double CudaElapsedUs(const Event& e) const {
+#ifdef PADDLE_WITH_CUDA
+    PADDLE_ENFORCE(e.has_cuda() && has_cuda());
+    PADDLE_ENFORCE(e.device() == device());
+    PADDLE_ENFORCE(cudaEventSynchronize(event_));
+    PADDLE_ENFORCE(cudaEventSynchronize(e.event()));
+    float ms;
+    PADDLE_ENFORCE(cudaEventElapsedTime(&ms, event_, e.event()));
+    return ms * 1000.0;
+#else
+    PADDLE_THROW("CUDA is not enabled");
+#endif
+  }
+
+ private:
+  EventKind kind_;
+  std::string name_;
+  uint32_t thread_id_;
+  int64_t cpu_ns_;
+  bool has_cuda_;
+#ifdef PADDLE_WITH_CUDA
+  cudaEvent_t event_ = nullptr;
+  int device_ = -1;
+#endif
+};
+
+struct EventList {
+  constexpr static std::size_t kMB = 1024 * 1024;
+  constexpr static std::size_t kEventBlockSize = 16 * kMB;
+  constexpr static std::size_t kEventSize = sizeof(Event);
+  constexpr static std::size_t kEventAlign = alignof(Event);
+  constexpr static std::size_t kNumBlock =
+      kEventBlockSize /
+      ((kEventSize + kEventAlign - 1) / kEventAlign * kEventAlign);
+
+  template <typename... Args>
+  void Record(Args&&... args) {
+    if (event_blocks.empty() || event_blocks.front().size() == kNumBlock) {
+      event_blocks.emplace_front();
+      event_blocks.front().reserve(kNumBlock);
+    }
+    event_blocks.front().emplace_back(std::forward<Args>(args)...);
+  }
+
+  std::vector<Event> Reduce() {
+    std::vector<Event> result;
+    for (auto& block : event_blocks) {
+      result.insert(result.begin(), std::make_move_iterator(block.begin()),
+                    std::make_move_iterator(block.end()));
+    }
+    event_blocks.clear();
+    return result;
+  }
+
+  std::forward_list<std::vector<Event>> event_blocks;
+};
+
+enum ProfilerState {
+  kDisabled,
+  kCPU,
+  kCUDA,
+};
+
+// The profiler state, the initial value is ProfilerState::kDisabled
+extern ProfilerState kState;
+// The global mutex
+extern std::mutex kAllEventListsMutex;
+// The total event lists of all threads
+extern std::list<std::shared_ptr<EventList>> kAllEventLists;
+// The thread local event list only can be accessed by the specific thread
+extern thread_local std::shared_ptr<EventList> kEventList;
+// The thread index of each thread
+extern thread_local int32_t kThreadId;
+// The kNextThreadId is a global counter for threads, by the kThreadId and
+// kNextThreadId, we can know how many threads have created EventList.
+extern uint32_t kNextThreadId;
+
+inline EventList& GetEventList() {
+  if (!kEventList) {
+    std::lock_guard<std::mutex> guard(kAllEventListsMutex);
+    kEventList = std::make_shared<EventList>();
+    kThreadId = kNextThreadId++;
+    kAllEventLists.emplace_front(kEventList);
+  }
+  return *kEventList;
+}
+
+inline void Mark(const std::string name,
```

Comment:
If you want to use RVO, you may not use `const`.

I suggest using `const T&` always. It is a simple rule.

Reply:
Done.